File: 2958.length-of-longest-subarray-with-at-most-k-frequency.py
#
# @lc app=leetcode id=2958 lang=python3
#
# [2958] Length of Longest Subarray With at Most K Frequency
#

# @lc code=start
from typing import List
from collections import Counter
class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        cnt = Counter()
        max_len = 0
        left, right = 0, 0

        while right < len(nums):
            x = nums[right]
            cnt[x] += 1
            right += 1

            while cnt[x] > k:
                cnt[nums[left]] -= 1
                left += 1

            max_len = max(max_len, right - left)
            
        return max_len
        
# @lc code=end

File: 2958.length-of-longest-subarray-with-at-most-k-frequency.cpp
/*
 * @lc app=leetcode id=2958 lang=cpp
 *
 * [2958] Length of Longest Subarray With at Most K Frequency
 */

// @lc code=start
#include<vector>
#include<iostream>
#include<algorithm>
#include<unordered_map>
using namespace std;

class Solution {
public:
    int maxSubarrayLength(vector<int>& nums, int k) {
        int max_len = 0;
        int left = 0, right = 0;
        unordered_map<int, int> freq; 

        for (right = 0; right < nums.size(); right++) {
            freq[nums[right]]++;
            while (freq[nums[right]] > k) {
                freq[nums[left]]--;
                left++;
            }
            max_len = max(max_len, right - left + 1);
        }
        return max_len;

    }
};
// @lc code=end

